### PR TITLE
Show more information in get_git_commit

### LINKED ISSUE #2457: Output more context when calling _timescaledb_internal.get_git_commit()
Currently, the output most likely yields the git tag of the commit, however for troubleshooting build issues, it would be good to include the sha also.

Suggested output:

```
tag text,
sha text not null,
committed timestamptz not null
```
We could also include if the repo was clean at the time of building, say, add output of `git status --porcelain=v2`, which would ensure tainted files would be known.

### PR DESCRIPTION
**Pull request by mkindahl**
Show more information in get_git_commit
  
The command `get_git_commit` shows quite a little information and
especially if mistakes are made it is hard to verify what the binary is
actually based on.

This commit extends the function to provide some more information:
specifically the SHA of the HEAD regardless of whether there is a tag
on it or not and the time for the HEAD commit.

```
postgres=# select * from _timescaledb_internal.get_git_commit();
        commit_tag        | commit_hash |      commit_time
--------------------------+-------------+------------------------
 1.7.4-10-g09b0b778-dirty | 09b0b778    | 2020-09-13 17:50:38+02
(1 row)
```

If git is not installed, the function `get_git_commit` will return an
error indicating that no git information is available. If some of the
fields are available, they will be emitted and the remaining fields
will be NULL.

Fixes #2457

### CODE REVIEW

**Review by pmwkaa — APPROVED** [review]
Looks good. Slightly concerned about `commit_sha` name maybe use `commit_id` or `commit_hash` name instead?

File: src/version.c [hunk 1]
```diff
@@ -20,21 +20,39 @@
 #define STR_EXPAND(x) #x
 #define STR(x) STR_EXPAND(x)
 
-static const char *git_commit = STR(EXT_GIT_COMMIT);
+static const char commit_sha[] = EXT_GIT_COMMIT_SHA;
```
Comment (erimatnor) [inline src/version.c:23]: Do we want to make this an exported string, e.g., `ts_commit_sha` so that one can use `strings` and `grep` directly on the binary?

Comment (mkindahl) [reply src/version.c:23]: I was considering this, but we might want to have some sort of tag so that we can easily find it, e.g., `sha:aec070645fe53ee3b3763059376134f058cc337247c978add178b6ccdfb0019f` and the same for the tag.

**Comment by mkindahl** [comment]
> Looks good. Slightly concerned about `commit_sha` name maybe use `commit_id` or `commit_hash` name instead?

It is referred to in the documentation as `sha1` or "SHA-1 checksum" so I think `commit_sha` is fine.

File: src/version.c (hunk 1 above)
Comment (mkindahl) [reply src/version.c]: I created three tags `commit-sha:`, `commit-tag:`, and `commit-time:` and added them. I do not think we need to export them since they are visible in the SO anyway as long as the link-time optimizer does not remove them. The prefix is removed before added to the result.

**Comment by svenklemm** [comment]
> Looks good. Slightly concerned about `commit_sha` name maybe use `commit_id` or `commit_hash` name instead?

I agree i think this should be named `commit_hash`

**Comment by svenklemm** [comment]
> > Looks good. Slightly concerned about `commit_sha` name maybe use `commit_id` or `commit_hash` name instead?
> 
> It is referred to in the documentation as `sha1` or "SHA-1 checksum" so I think `commit_sha` is fine.

git is planning to move away from sha1 so a more generic name makes sense https://git-scm.com/docs/hash-function-transition/

**Comment by mkindahl** [comment]
> > > Looks good. Slightly concerned about `commit_sha` name maybe use `commit_id` or `commit_hash` name instead?
> > 
> > 
> > It is referred to in the documentation as `sha1` or "SHA-1 checksum" so I think `commit_sha` is fine.
> 
> git is planning to move away from sha1 so a more generic name makes sense https://git-scm.com/docs/hash-function-transition/

Wasn't aware of that. Makes sense to call it `hash` then.

File: src/version.c [hunk 2]
```diff
@@ -10,31 +10,54 @@
...
 Datum
 ts_get_git_commit(PG_FUNCTION_ARGS)
 {
-	size_t var_size = VARHDRSZ + strlen(git_commit);
-	text *version_text = (text *) palloc(var_size);
+	TupleDesc tupdesc;
+	HeapTuple tuple;
+	Datum values[3];
+	bool nulls[3] = { false };
+	const char *const tag_ptr = strchr(commit_tag, ':');
```
Comment (erimatnor) [inline src/version.c:36]: I don't understand why we first create a const char string for each commit information with their unique prefix, just to remove the prefix here.

If the idea is of the prefixes is to be able to find them in the binary (without calling this function), why not just directly use `EXT_GIT_COMMIT_TAG` here without having to strip the prefix that was just added above?

File: CMakeLists.txt [hunk 3]
```diff
@@ -164,7 +164,7 @@ if (NOT PG_CONFIG)
   message(FATAL_ERROR "Unable to find 'pg_config'")
 endif ()
 
-find_package(Git)
+find_package(Git REQUIRED)
```
Comment (erimatnor) [inline CMakeLists.txt:167]: Unclear if this is a great idea for windows builds. Then we are adding Git as a dependency for building on all platforms.

File: src/version.c (hunk 2 above)
Comment (mkindahl) [reply src/version.c:36]: Mainly because I want to ensure that the static variable is referenced so that it is not removed by an optimizing linker.

File: src/version.c [hunk 4]
```diff
@@ -10,31 +10,54 @@
...
 #include "version.h"
 #include "config.h"
 
-#define STR_EXPAND(x) #x
-#define STR(x) STR_EXPAND(x)
-
-static const char *git_commit = STR(EXT_GIT_COMMIT);
+/* Export the strings to that we can read them using strings(1). We add a
+ * prefix so that we can easily find it using grep(1) */
+static const char commit_hash[] = "commit-hash:" EXT_GIT_COMMIT_HASH;
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/version.c:23]: There were a discussion about regarding this. They are going to call it "hash" going forward: https://git-scm.com/docs/hash-function-transition/

File: CMakeLists.txt (hunk 3 above)
Comment (mkindahl) [reply CMakeLists.txt:167]: Since this function is about providing git information, we already have that (it passes the Windows build in the CI). An option would be to make Git optional but then require building from a source tar, which would include the generated file with git information, or Git, and not make the function available otherwise.

File: src/version.c (hunk 2 above)
Comment (erimatnor) [reply src/version.c:36]: Isn't it better to use a compiler pragma/attribute on the static variable instead to make it explicit that this variable should not be used.

The intention of this unnecessary parsing is non-obvious and it is easy for someone in the future to not understand its purpose and thus removing this code.

Comment (mkindahl) [reply src/version.c:36]: Pragmas are usually not as portable, but I can switch.

File: CMakeLists.txt (hunk 3 above)
Comment (erimatnor) [reply CMakeLists.txt:167]: I am not sure I am following. AFAIK, Git was not required previously---at least that was not the intention. That things pass _our_ CI is a different matter since we control that environment and can install whatever dependencies we like. But I am not sure we should require all those dependencies for other source builds. Compiling from a tar ball is one example.

File: src/version.c (hunk 2 above)
Comment (erimatnor) [reply src/version.c]: Yes, I realize that. But we use them for other things already, typically by defining a portable macro for the platforms we support. Usually, Windows is the odd-one out as GCC and Clang are often compatible.

File: src/gitcommit.h.in [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+ */
+
+#ifndef GITCOMMIT_H_
+#define GITCOMMIT_H_
+
+#cmakedefine EXT_GIT_COMMIT_TAG "@EXT_GIT_COMMIT_TAG@"
+#cmakedefine EXT_GIT_COMMIT_HASH "@EXT_GIT_COMMIT_HASH@"
+#cmakedefine EXT_GIT_COMMIT_TIME "@EXT_GIT_COMMIT_TIME@"
+
+#endif // GITCOMMIT_H_
```
Comment (erimatnor) [inline src/gitcommit.h.in:14]:
```suggestion
#endif /* GITCOMMIT_H_ */
```

File: CMakeLists.txt (hunk 3 above)
Comment (mkindahl) [reply CMakeLists.txt:167]: Right now, the function is defined to return something else than the git version if git does not exists, which I think is wrong since we use this for debugging we do not get usable information if the functions invent a commit tag that it is built from (the project version).

I added the requirements since it seemed the intention was to always provide that function, but a better option would be to remove the function if git is not installed and not require it Git when building from a non-git source.

When building from a source tarball, configuration files are usually included and since the git information is in a configuration file, that would be included in the source tarball. However, it seems we have not configured for building source tarballs, which would explain that the function uses the project version instead of the git tag when git is not installed.

I would suggest the following then (for this PR):

- Do not require Git.
- If Git is not installed, do not provide `get_git_commit`. This will make it clear that the extension is built is from a non-git source build. The extension version is still available in `\dx` so we can get that information regardless.
- If Git is installed, provide the `get_git_commit`. This will make it clear that it is built from a git build and what git commit that the extension is built from, regardless of whether it is a Docker image, a package, or a source build in a git repository.

File: src/version.c (hunk 2 above)
Comment (mkindahl) [reply src/version.c]: Added a new annotation and use that instead.

File: src/gitcommit.h.in (hunk 5 above)
Comment (mkindahl) [reply src/gitcommit.h.in]: Fixed.

File: CMakeLists.txt (hunk 3 above)
Comment (erimatnor) [reply CMakeLists.txt:167]: My only concern is that we should support tar-ball builds. Even if Git is installed when you build tar-balls, Git won't provide useful information since you are not in a repo. Thus, requiring GIt seems wrong. We should fall back to other information sources, like reading the version from `version.config` (perhaps with a useful suffix) or similar.

I think it is fine to do best effort w.r.t. version info, but for "released" tar balls we should be able to include a pre-generated version file. Note, however, that, e.g., GitHub provides tar-ball downloads automatically by simply archiving the Git repo: https://github.com/timescale/timescaledb/archive/2.0.0-rc3.tar.gz. Such a tar ball won't include a version.h file and in that case I think we must fall back to`version.config` as outlined above.

I am not sure what you mean by "not provide `get_git_commit`"--- I think this function will exist regardless, although the returned info can of course vary.

Comment (mkindahl) [reply CMakeLists.txt]: Changed the PR to not require Git and not define the function if the git information cannot be retrieved.

File: src/version.c [hunk 6]
```diff
@@ -10,32 +10,53 @@
...
+#include "annotations.h"
 #include "gitcommit.h"
 #include "version.h"
 #include "config.h"
 
-#define STR_EXPAND(x) #x
-#define STR(x) STR_EXPAND(x)
-
-static const char *git_commit = STR(EXT_GIT_COMMIT);
+#ifdef GIT_FOUND
```
Comment (erimatnor) [inline src/version.c:22]: I think we need to provide this implementation even if the git information is not available. Otherwise the SQL function will have a missing C-implementation. (Also note that tests don't pass, but I don't know if it is for this reason or not).

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -72,31 +72,47 @@ if (CMAKE_BUILD_TYPE MATCHES Debug)
...
-else ()
-  add_custom_command(
-    OUTPUT ${GITCOMMIT_H}
-    WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
-    COMMAND ${CMAKE_COMMAND} -E echo_append "#define EXT_GIT_COMMIT " > ${GITCOMMIT_H}
-    COMMAND sh -c "if `${GIT_EXECUTABLE} status > /dev/null 2>&1`; then ${GIT_EXECUTABLE} describe --abbrev=4 --dirty --always --tags ; else echo ${PROJECT_VERSION_MOD} ; fi" >> ${GITCOMMIT_H}
-    COMMENT "Generating gitcommit.h"
-    VERBATIM)
-endif (WIN32)
+if(GIT_FOUND)
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/CMakeLists.txt:76]: There were an explanation in the PR, albeit terse. As we discussed, the code is now updated to return an explicit error instead of just erroring with a "function not defined", so I updated the PR comment and the commit description to give a more elaborate description.

File: src/version.c (hunk 6 above)
Comment (mkindahl) [reply src/version.c]: In the previous commit, the function definition was excluded if Git was not found, so this worked as expected, but after our discussions it now returns an explicit error, hence it needs to be always defined.

File: .github/workflows/update-test.yaml [hunk 8]
```diff
@@ -29,7 +29,7 @@ jobs:
 
     - name: Update tests ${{ matrix.pg }}
       run: |
-        ./scripts/test_updates_pg${{ matrix.pg_major}}.sh
+        bash -x ./scripts/test_updates_pg${{ matrix.pg_major}}.sh
```
Comment (erimatnor) [inline .github/workflows/update-test.yaml:32]: Nit, use `sh` instead since the shell might not be bash.

File: .github/workflows/alpine-32bit-build-and-test.yaml [hunk 9]
```diff
@@ -30,7 +30,7 @@ jobs:
 
     - name: Install build dependencies
       run: |
-        apk add --no-cache --virtual .build-deps coreutils dpkg-dev findutils gcc libc-dev make util-linux-dev diffutils cmake openssl-dev sudo flex bison
+        apk add --no-cache --virtual .build-deps coreutils dpkg-dev findutils gcc libc-dev make util-linux-dev diffutils cmake openssl-dev sudo flex bison git
```
Comment (erimatnor) [inline .github/workflows/alpine-32bit-build-and-test.yaml:33]: Do we need this dependency now that Git is still optional? Isn't it good to also run some tests where we don't have Git installed and build the dummy version of the get_git_commit function?

File: scripts/docker-build.sh [hunk 10]
```diff
@@ -54,7 +54,7 @@ create_postgres_build_image() {
     docker run -d --name ${BUILD_CONTAINER_NAME} --env POSTGRES_HOST_AUTH_METHOD=trust -v ${BASE_DIR}:/src postgres:${PG_IMAGE_TAG}
 
     # Install build dependencies
-    docker exec -u root ${BUILD_CONTAINER_NAME} /bin/bash -c "apk add --no-cache --virtual .build-deps gdb git coreutils dpkg-dev gcc libc-dev make cmake util-linux-dev diffutils openssl-dev && mkdir -p /build/debug"
+    docker exec -u root ${BUILD_CONTAINER_NAME} /bin/bash -c "apk add --no-cache --virtual .build-deps gdb coreutils dpkg-dev gcc git libc-dev make cmake util-linux-dev diffutils openssl-dev && mkdir -p /build"
```
Comment (erimatnor) [inline scripts/docker-build.sh:57]: Same question as above, regarding need to install git.

File: scripts/docker-build.sh [hunk 11]
```diff
@@ -74,10 +74,9 @@ build_timescaledb()
...
 
     run_postgres_build_image ${PG_IMAGE}
-
+    
     # Build and install the extension with debug symbols and assertions
-    tar -c -C ${BASE_DIR} {src,sql,test,scripts,tsl,version.config,CMakeLists.txt,timescaledb.control.in} | docker cp - ${BUILD_CONTAINER_NAME}:/build/
-    docker exec -u root ${BUILD_CONTAINER_NAME} /bin/bash -c "cd /build/debug && cmake -DUSE_OPENSSL=${USE_OPENSSL} -DREGRESS_CHECKS=OFF -DCMAKE_BUILD_TYPE=${BUILD_TYPE} .. && make && make install && echo \"shared_preload_libraries = 'timescaledb'\" >> /usr/local/share/postgresql/postgresql.conf.sample && echo \"timescaledb.telemetry_level=off\" >> /usr/local/share/postgresql/postgresql.conf.sample && cd / && rm -rf /build"
+    docker exec -u root ${BUILD_CONTAINER_NAME} /bin/bash -c "cd /build && cmake -DUSE_OPENSSL=${USE_OPENSSL} -DREGRESS_CHECKS=OFF -DCMAKE_BUILD_TYPE=${BUILD_TYPE} /src && make && make install && echo \"shared_preload_libraries = 'timescaledb'\" >> /usr/local/share/postgresql/postgresql.conf.sample && echo \"timescaledb.telemetry_level=off\" >> /usr/local/share/postgresql/postgresql.conf.sample && cd / && rm -rf /build"
```
Comment (erimatnor) [inline scripts/docker-build.sh:79]: I don't understand this change, in particular the removal the removal of tar and copy. This seems unrelated to the main changes in this PR. At least an explanation is in order.

File: src/version.c [hunk 12]
```diff
@@ -10,32 +10,88 @@
...
 #include "gitcommit.h"
 #include "version.h"
 #include "config.h"
 
-#define STR_EXPAND(x) #x
-#define STR(x) STR_EXPAND(x)
+/* Export the strings to that we can read them using strings(1). We add a
+ * prefix so that we can easily find it using grep(1). We only bother about
+ * generating them if the relevant symbol is defined. */
+#ifdef EXT_GIT_COMMIT_HASH
```
Comment (erimatnor) [inline src/version.c:25]: Will this work as expected?

Even if `@EXT_GIT_COMMIT_TAG@` is empty in gitcommit.h.in, I suspect  `EXT_GIT_COMMIT_TAG` will still be defined (albeit empty)?

File: test/expected/version.out [hunk 13]
```diff
@@ -5,7 +5,7 @@
 select pg_typeof(git) from _timescaledb_internal.get_git_commit() AS git;
```
Comment (erimatnor) [inline test/expected/version.out:5]: I suggest changing this to typeof each returned column. Or at least add an additional line to the test to do that, similar to get_os_info below. Will it work for NULL values?

File: src/version.c [hunk 14]
```diff
@@ -10,32 +10,88 @@
...
+#endif
+
+#ifdef EXT_GIT_COMMIT_HASH
+	values[1] = CStringGetTextDatum(EXT_GIT_COMMIT_HASH);
+#else
+	nulls[1] = true;
+#endif
+
+#ifdef EXT_GIT_COMMIT_TIME
+	values[2] = DirectFunctionCall3(timestamptz_in,
```
Comment (erimatnor) [inline src/version.c:74]: Do we know that the the Git timestamp is always a parsable TimestampTz string?

**Review by erimatnor — APPROVED** [review]
Approving, but there are some things that I believe needs an extra look before merging.

File: scripts/docker-build.sh [hunk 15]
```diff
@@ -54,7 +54,7 @@ create_postgres_build_image() {
     docker run -d --name ${BUILD_CONTAINER_NAME} --env POSTGRES_HOST_AUTH_METHOD=trust -v ${BASE_DIR}:/src postgres:${PG_IMAGE_TAG}
 
     # Install build dependencies
-    docker exec -u root ${BUILD_CONTAINER_NAME} /bin/bash -c "apk add --no-cache --virtual .build-deps gdb git coreutils dpkg-dev gcc libc-dev make cmake util-linux-dev diffutils openssl-dev && mkdir -p /build/debug"
+    docker exec -u root ${BUILD_CONTAINER_NAME} /bin/bash -c "apk add --no-cache --virtual .build-deps gdb coreutils dpkg-dev gcc git libc-dev make cmake util-linux-dev diffutils openssl-dev && mkdir -p /build/debug"
```
Comment (erimatnor) [inline scripts/docker-build.sh:57]: Looks like `git` is just moved to later in the list. Looks like this change isn't strictly needed in that case.

File: src/version.c (hunk 12 above)
Comment (erimatnor) [reply src/version.c:25]: Still unclear what happens if `EXT_GIT_COMMIT_HASH` is defined like this:

```
#define EXT_GIT_COMMIT_HASH
```

File: .github/workflows/alpine-32bit-build-and-test.yaml (hunk 9 above)
Comment (mkindahl) [reply .github/workflows/alpine-32bit-build-and-test.yaml:33]: I think we should require git also for this case. Before, the tests was designed to not fail when Git was not available, but that is not the case right now and it will fail if git is not available (because the function generate an error rather than some output).

Since this is an internal function intended for debugging, I do not think a full test is necessary, but if you feel strongly about testing this, we can create a separate test that is only added if Git is absent.

File: src/version.c (hunk 12 above)
Comment (mkindahl) [reply src/version.c:25]: But it is not defined like that in the generated file, it is either

```
#define EXT_GIT_COMMIT_HASH "2beeb3fc"
```

or 

```
/* #undef EXT_GIT_COMMIT_HASH */
```

If a builder decides on editing the file and putting anything else there, they can break the build whatever we do, so there is no reason to handle that case.

File: scripts/docker-build.sh (hunk 15 above)
Comment (mkindahl) [reply scripts/docker-build.sh:57]: Correct. I'll remove it if I make more changes to the PR.

File: src/version.c (hunk 12 above)
Comment (erimatnor) [reply src/version.c:25]: That's what I wanted to know.
